tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -20,8 +20,19 @@
 K_SEM_DEFINE(uthread_start_sem, 0, 1);
 K_SEM_DEFINE(uthread_end_sem, 0, 1);
 K_SEM_DEFINE(test_revoke_sem, 0, 1);
+K_SEM_DEFINE(expect_fault_sem, 0, 1);
 
 static bool give_uthread_end_sem;
+static bool expect_fault;
+
+/*
+ * We need something that can act as a memory barrier
+ * from usermode threads to ensure expect_fault has
+ * been updated.  We'll just make an arbitrary system
+ * call to force one.
+ */
+#define BARRIER() k_sem_give(&expect_fault_sem)
```

Comment:
can you just use the compiler_barrier() macro defined by toolchain.h?

Reply:
i'm not sure the compiler barrier contains what is needed (at least for ARM)

Reply:
please elaborate.

Reply:
We need more than just a compiler barrier here; we need to ensure that the write to expect_fault has completed before the potentially faulting instruction, so we need the processor to honor the memory ordering as well. If this code was running in supervisor mode, I'd just do a __DSB() or __DMB(). The easiest way to do so from usermode seems to be to perform a system call, and repeatedly calling k_sem_give on this semaphore serves that purpose.

Reply:
gotcha.
more of an "execution barrier" than a "memory barrier".
it's hard to get the compiler to do this :)